ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java

```diff
@@ -0,0 +1,235 @@
+   * be done during normal instance construction without letting a reference to an incomplete constructed instance escape.
+   */
+  void init() {
+    if (!initialized.getAndSet(true)) {
+      if (scheduler != null) {
+        Runnable updater = () -> {
+          try {
+            processAccountJsonFile();
+          } catch (Exception e) {
+            logger.error("Exception occurred when loading JSON account data", e);
+            accountServiceMetrics.fetchRemoteAccountErrorCount.inc();
+          }
+        };
+
+        int initialDelay = new Random().nextInt(accountConfig.updaterPollingIntervalMs + 1);
+        scheduler.scheduleAtFixedRate(updater, initialDelay, accountConfig.updaterPollingIntervalMs, TimeUnit.MILLISECONDS);
+        logger.info(
+            "Background account updater will fetch accounts from remote starting {} ms from now and repeat with interval={} ms",
+            initialDelay, accountConfig.updaterPollingIntervalMs);
+      } else {
+        processAccountJsonFile();
+      }
+    }
+  }
+
+  @Override
+  protected void checkOpen() {
+  }
+
+  /**
+   * Since the JSON file account service gets all it's data from a local JSON file it can't update account data
+   * for all nodes. Therefor calling this method will always throw an {@link UnsupportedOperationException}.
+   *
+   * @param accounts The collection of {@link Account}s to update. Cannot be {@code null}.
+   * @return
+   * @throws UnsupportedOperationException Always thrown when this method is invoked.
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
+    throw new UnsupportedOperationException(
+        "Service does not support updating of accountsById via API. Modify account JSON files on all nodes instead.");
+  }
+
+  /**
+   * Makes the service release all its resources. Be aware that the service is no longer usable after this. This
+   * instance can't be restarted and will need to be re-instantiated.
+   */
+  @Override
+  public void close() throws IOException {
+    if (watchAccountFileThread != null && !watchAccountFileThread.isInterrupted()) {
+      watchAccountFileThread.interrupt();
+    }
+  }
+
+  /** How often we check if the JSON account file has been modified in milliseconds. */
+  private static int fileWwatchIntervalInMs = 10 * 1000;
+  /** Reference to the thread which checks the account file for changes at regular intervals. Will stop gracefully
+   * when an interrupt is send to it.*/
+  private volatile Thread watchAccountFileThread;
+
+  /**
+   * Starts a thread which checks (at fixed intervals) if the JSON account file has changed. If it has it triggers
+   * and update of the accounts.
+   */
+  private void watchJsonFile() {
+    watchAccountFileThread = new Thread(() -> {
+      logger.info("Starting JSON account file monitor. File check interval is {}ms", fileWwatchIntervalInMs);
+
+      while (!Thread.interrupted()) {
```

Comment:
Have you looked into using WatchService to react to file changes instead of using a polling interval: https://docs.oracle.com/javase/tutorial/essential/io/notification.html

Reply:
Ah, it actually seems like this method is not used, since it is handled by the updater `ScheduledExecutorService`, so this method could probably be removed

Reply:
I did consider using the `WatchService`. I can understand the the question so I have added the following comment to the class:

> This implementation uses a thread to check for changes in the JSON file at scheduled intervals. While using a {@link java.nio.file.WatchService} might seem more intuitive the docs of the WatchService specify that the characteristics of the WatchService are highly platform (Linux, Windows, Solaris, etc.) dependent. Going as far as specifying events may under certain conditions not be fired. Therefor having a fallback which at intervals checks if the JSON file has changed will probably always be needed to ensure no changes are missed even if a WatchService would be added.

Adding a `WatchService` to the implementation could be used to improve the change detection. But I don't think that one can do with the regular intervals checks to ensure all changes are noticed.